Add a bunch of missing class codes

dylanmckay opened the pull request:
Previously the class code parser only understood the four main class
codes.

This modified it so that it understands all class codes.

After this, it is possible to parse mDNS records from a ChromeCast.

There is a table of possible values listed on this page
http://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml

After this is added, would you please be able to release a new version to crates.io :)

tailhook commented:
Well, I'm not sure about forward compatibility of this approach. Could we just add mdns class?

The problem with your approach is that if we have something now that fits unassigned class and you have a `match` for that value, in next version it may become an enum and you will miss that values.

The old approach is not strictly compatible too, as adding an option to public enum isn't considered backward compatible too, but at least it's a simple to fix compilation error not just silently missing event.

dylanmckay commented:
> Could we just add mdns class?

We could do that, but none of this stuff is mDNS specific. These are all codes that can be send from a DNS server in the future and be completely valid.

The backwards compatibility thing is a good argument though. One upside of my approach is that it will force users to update their code to handle new class types (through compilation errors), whereas as it currently stands it will just crash instead. Alongside that, the last addition to the 'class codes' was an RFC from 2013 (which is now almost 4 years ago). Before that, the last modification was in **1997**.

tailhook commented:
> whereas as it currently stands it will just crash instead

By "crash" you mean it treats packet as invalid, right? If it panics it's definitely a bug.

> Alongside that, the last addition to the 'class codes' was an RFC from 2013 (which is now almost 4 years ago). Before that, the last modification was in 1997.

This means we surely can keep up with the changes in the protocol :)

----

Another question is: you've added `ANY` and `NONE` to Class but according to docs they are `QCLASS`, meaning that they can only be used in queries not in responses. We already have `Any` in `QueryClass` by the way. Was that intentional?

dylanmckay commented:
> By "crash" you mean it treats packet as invalid, right? If it panics it's definitely a bug.

dylanmckay commented:
> By "crash" you mean it treats packet as invalid, right? If it panics it's definitely a bug.

Yes, I do mean that. It depends on how the clients handle the `Error`.

> Another question is: you've added ANY and NONE to Class but according to docs they are QCLASS, meaning that they can only be used in queries not in responses. We already have Any in QueryClass by the way. Was that intentional?

It was unintentional, that makes sense though. I can rename them to `QueryNone` and `QueryAny` if that's preferable.

tailhook commented:
> Yes, I do mean that. It depends on how the clients handle the Error.

Okay. You shouldn't rely on packet to be always right. You can receive any kind of garbage from the network at any time. This is especially important if you use UDP (coparing to TCP). So ignoring a packet is a better strategy than crashing on bad packet at all times. (I.e. some garbage can arrive earlier than reply to your request).

> It was unintentional, that makes sense though. I can rename them to QueryNone and QueryAny if that's preferable.

There is a separate `enum QueryClass` so I expect you add None class there :)

So do you have dump of a packet that you can't parse? It would be good to add it to the tests.

And I'm still not sure, will it suit your use case if we just add None and keep `InvalidClass` on unassigned and reserved classes?

dylanmckay commented:
For my personal use case, I am receiving mDNS responses from a Chromecast with a class code of `32769`. This corresponds to one of the unassigned class numbers.

Multicast DNS is defined in [RFC 6762](https://tools.ietf.org/html/rfc6762).

It has this to say

> 18.13.  Repurposing of Top Bit of rrclass in Resource Record Sections
> 
>    In the Resource Record Sections of a Multicast DNS response, the top
>    bit of the rrclass field is used to indicate that the record is a
>    member of a unique RRSet, and the entire RRSet has been sent together
>    (in the same packet, or in consecutive packets if there are too many
>    records to fit in a single packet).

So it looks like the class field for mDNS can actually take _any_ value. The only way to reliably support mDNS in this library is to get rid of the `InvalidClass` error and return the raw code. Is that something you're okay with?

tailhook commented:
> So it looks like the class field for mDNS can actually take any value. The only way to reliably support mDNS in this library is to get rid of the InvalidClass error and return the raw code. Is that something you're okay with?

Wow! Yes, then it's fine. mDNS case is an important one.

The only question left: if we still want the `Class` and `QueryClass` split. Previously we kept it to reject ANY value in responses. But it doesn't seem reasonable now. Should we merge them?

dylanmckay commented:
I think we should leave them, as there is a difference between them. It's still doesn't make sense to have `ANY` or `NONE` inside a resource record.

dylanmckay commented:
I've reworked this to add a `Class::Other { code: u16 }` variant.

tailhook commented:
Is this unassigned class used only in responses but never in queries? Or mdns have no notion of queries at all?. I'm not too familiar with the protocol (will read details when have time, just trying to get thing thing merged faster)

(sorry for discussing it too much, but it's important to get details right, because we have to live with them for years...)

dylanmckay commented:
An excerpt from the RFC

> 18.13. Repurposing of Top Bit of rrclass in Resource Record Sections

This shows that the unassigned numbers are indeed only used in responses.

> (sorry for discussing it too much, but it's important to get details right, because we have to live with them for years...)

No problems here :)

tailhook commented:
Well, sorry I haven't noticed it before but:

> 18.13. Repurposing of Top Bit of rrclass in Resource Record Sections

It means only top bit is different. And your value `32769` is top bit set plus normal `IN` class.

So what we have to do: extract the top bit into separate flag and use class as it's currently defined. (I can't make a good name for flag off the top of head now)

It looks like there is the same issue with query class:

> 18.12.  Repurposing of Top Bit of qclass in Question Section

Does it make sense?

dylanmckay commented:
That sounds like a really nice design

Commit:
Teach DNS parser about specific mDNS class bits

This adds support for two mDNS special cases.

The top bits of both Class and QueryClass have special meaning.

For query classes, the top bit indicates whether or not unicast
responses are preferred.

For resource record classes, the top bit indicates whether or not there
are more resource records in future packets.

dylanmckay commented:
I've ripped out the changes to the `Class` and `QueryClass` enums and added flags to the structs to handle the top bit special cases.

tailhook commented:
Looks good. Thanks!

tailhook commented:
`v0.6.0`
